crates/relayer/src/chain/axon/monitor.rs

```diff
@@ -89,22 +81,67 @@ impl AxonEventMonitor {
         fields(chain = %self.chain_id)
     )]
     pub fn run(mut self) {
-        debug!("starting Axon event monitor");
-        let rt = self.rt.clone();
-        rt.block_on(async {
+        if let Next::Continue = self.update_subscribe(false) {
+            info!("start Axon event monitor for {}", self.chain_id);
             loop {
-                match self.run_loop().await {
-                    Next::Continue => continue,
-                    Next::Abort => break,
+                if let Next::Abort = self.run_loop() {
+                    break;
                 }
             }
-        });
-        debug!("event monitor is shutting down");
+            debug!("event monitor is shutting down");
+        }
         // TODO: close client
     }
 
-    fn update_subscribe(&mut self) -> Next {
-        if let Ok(cmd) = self.rx_cmd.try_recv() {
+    pub fn restore_event_tx_hashes(
+        &self,
+        latest_block_count: u64,
+    ) -> Result<Vec<IbcEventWithHeight>> {
+        let contract = Arc::new(Contract::new(
+            self.contract_address,
+            Arc::clone(&self.client),
+        ));
+        let restore_block_number = self
+            .start_block_number
+            .checked_sub(latest_block_count)
+            .ok_or(Error::others(format!(
+                "latest_block_count {latest_block_count} exceeds start_block_number {}",
+                self.start_block_number
+            )))?;
+        let events = self
+            .rt
+            .block_on(
+                contract
+                    .events()
+                    .from_block(restore_block_number)
+                    .to_block(self.start_block_number)
+                    .query_with_meta(),
+            )
+            .map_err(|e| Error::others(e.to_string()))?
+            .into_iter()
+            .map(|(event, meta)| {
+                IbcEventWithHeight::new_with_tx_hash(
+                    event.into(),
+                    Height::from_noncosmos_height(meta.block_number.as_u64()),
+                    meta.transaction_hash.into(),
+                )
+            })
+            .collect::<Vec<_>>();
+        debug!(
+            "restored {} events on contract {}",
+            events.len(),
+            self.contract_address
+        );
+        Ok(events)
+    }
+
+    fn update_subscribe(&mut self, use_try: bool) -> Next {
+        let result = if use_try {
+            self.rx_cmd.try_recv().ok()
+        } else {
+            self.rx_cmd.recv().ok()
```

Comment:
When the sender end is closed, this will always return `Continue`? (I guess the sender is supposed to send `Shutdown` before dropping, but we can have more robust error handling here, and `AxonChain` is not actually doing this.)

Reply:
I guess this channel cannot be closed, because monitoring loop should be alive for checking IBC events, and Axon endpoint uses different implementation of this, which is supposed to be the same way of manipulating and should be fixed I guess

Reply:
OK let's not block on this. We can add some more robust error handling later.